spyder/plugins/preferences/plugin.py

```diff
@@ -0,0 +1,169 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright © Spyder Project Contributors
+# Licensed under the terms of the MIT License
+# (see spyder/__init__.py for details)
+
+"""
+Spyder preferences plugin.
+
+This plugin is on charge of managing preference pages and tabs for all plugins
+in Spyder, both internal and external.
+"""
+
+# Standard library imports
+import logging
+from typing import Union
+
+# Third-party imports
+from qtpy.QtGui import QIcon
+
+# Local imports
+from spyder.api.plugins import SpyderPluginV2, SpyderPlugin
+from spyder.config.base import _
+from spyder.config.user import NoDefault
+from spyder.plugins.preferences.api import PreferencePages, GeneralConfigPage
+from spyder.plugins.preferences.general import MainConfigPage
+from spyder.plugins.preferences.widgets.container import PreferencesContainer
+
+logger = logging.getLogger(__name__)
+
+BaseType = Union[int, float, bool, complex, str, bytes]
+IterableType = Union[list, tuple]
+BasicType = Union[BaseType, IterableType]
+
+
+class Preferences(SpyderPluginV2):
+    """
+    Spyder preferences plugin.
+
+    This class manages all the preference pages and tabs for all internal
+    and external plugins, as well enabling other plugins to add configurations
+    to other sections.
+    """
+
+    NAME = 'preferences'
+    CONF_SECTION = 'preferences'
+    CONF_FILE = False
+    CONTAINER_CLASS = PreferencesContainer
+
+    NEW_API = 'new'
+    OLD_API = 'old'
+
+    def __init__(self, parent, configuration=None):
+        super().__init__(parent, configuration)
+        self.config_pages = {
+            PreferencePages.General: (self.NEW_API,
+                                      lambda plugin, dlg: MainConfigPage(
+                                          dlg, plugin),
+                                      self.get_main())
+        }
+        self.config_tabs = {}
+
+    def register_plugin_preferences(
+            self, plugin: Union[SpyderPluginV2, SpyderPlugin]) -> None:
+        if (hasattr(plugin, 'CONF_WIDGET_CLASS') and
+                plugin.CONF_WIDGET_CLASS is not None):
+            # New API
+            Widget = plugin.CONF_WIDGET_CLASS
+            if issubclass(plugin.CONF_WIDGET_CLASS, GeneralConfigPage):
+                Widget = lambda plugin, dlg: plugin.CONF_WIDGET_CLASS(
+                    dlg, plugin)
+
+            self.config_pages[plugin.NAME] = (self.NEW_API, Widget, plugin)
+
+            # Check if the plugin adds new configuration options to other
+            # sections
+            if plugin.ADDITIONAL_CONF_OPTIONS is not None:
```

Comment:
I see one important problem with adding options to another plugin: it won't be possible to change their defaults by bumping `CONF_VERSION` (as we do right now).

Have you considered that case?